Sources/Machine/abstract_machine.cc

```diff
@@ -12,20 +12,106 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "abstract_machine.hpp"
+#include "Machine/abstract_machine.hpp"
+
+#include <mpi.h>
+#include <nlohmann/json.hpp>
 
 #include <fstream>
 
-#include <nlohmann/json.hpp>
+#include "Utils/mpi_interface.hpp"
+#include "Utils/random_utils.hpp"
 
 namespace netket {
 
+namespace {
+inline unsigned ReallyRandomSeed() {
+  std::random_device rd;
+  return rd();
+}
+}  // namespace
+
+void AbstractMachine::InitRandomPars(double sigma,
+                                     nonstd::optional<unsigned> seed) {
+  VectorType parameters(Npar());
+  const auto comm = MPI_COMM_WORLD;
+  constexpr auto root = 0;
+  int rank;
+  MPI_Comm_rank(comm, &rank);
+  if (rank == root) {
+    default_random_engine generator{seed.has_value() ? *seed
+                                                     : ReallyRandomSeed()};
+    std::generate(parameters.data(), parameters.data() + parameters.size(),
+                  [&generator, sigma]() {
+                    std::normal_distribution<double> dist(0.0, sigma);
+                    return Complex{dist(generator), dist(generator)};
+                  });
+  }
+  auto status = MPI_Bcast(parameters.data(), parameters.size(),
+                          MPI_DOUBLE_COMPLEX, root, comm);
+  if (status != MPI_SUCCESS) throw MPIError{status, "MPI_Bcast"};
+  SetParameters(parameters);
+}
+
+void AbstractMachine::LogVal(Eigen::Ref<const RealRowMatrixType> v,
+                             Eigen::Ref<VectorType> out,
+                             const any & /*unused*/) {
+  if (v.cols() != Nvisible()) {
+    std::ostringstream msg;
+    msg << __FUNCTION__ << ": invalid dimension: [" << v.rows() << ", "
+        << v.cols() << "]; expected [?, " << Nvisible() << "]";
+    throw InvalidInputError{msg.str()};
+  }
+  if (v.rows() != out.size()) {
+    std::ostringstream msg;
+    msg << __FUNCTION__ << ": input dimension [" << v.rows() << ", " << v.cols()
+        << "] and output dimension [" << out.size() << "] are incompatible";
+    throw InvalidInputError{msg.str()};
+  }
+  for (auto i = Index{0}; i < v.rows(); ++i) {
+    out(i) = LogValSingle(v.row(i));
+  }
+}
+
+AbstractMachine::VectorType AbstractMachine::LogVal(
+    Eigen::Ref<const RealRowMatrixType> v, const any &cache) {
+  VectorType out(v.rows());
+  LogVal(v, out, cache);
+  return out;
+}
+
+void AbstractMachine::DerLog(Eigen::Ref<const RealRowMatrixType> v,
+                             Eigen::Ref<RowMatrixType> out, const any &cache) {
+  if (v.cols() != Npar()) {
```

Comment:
The error checks here are not correct. For example, `v.cols()` should probably be `== Nvisible()`, not `Npar()`.